README.md
# 助数詞を練習: Let's Review Japanese Counters!

[![CircleCI](https://circleci.com/gh/ahlec/jyosuushi.svg?style=svg)](https://circleci.com/gh/ahlec/jyosuushi)
[![Twitter Follow](https://img.shields.io/twitter/follow/Jyosuushi.svg?style=social)](https://twitter.com/Jyosuushi)

A browser-based webapp to facilitate studying [Japanese counters](https://en.wikipedia.org/wiki/Japanese_counter_word). A lightweight app inspired by the simplicity of [WaniKani](https://www.wanikani.com/), this application allows the user to select packs of counters they would like to study, and then presents a quiz with random amounts of random items, requiring the user to correctly count the specified item in Japanese.

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and deployment.

### Prerequisites

The project uses [Yarn](https://yarnpkg.com/) for package management.

Additionally, if you're going to be changing the favicon, make sure to globally install `cli-real-favicon`.

```
yarn global add cli-real-favicon
```

I haven't found a way to get this to work without being a global installation.

### Installing

1. Clone the repo

```
git clone https://github.com/ahlec/jyosuushi.git
```

2. Install packages

```
yarn install
```

3. Start the development server

```
yarn start
```

At this point, you can navigate to `http://localhost:8080/`. As of right now, there is no persistence of data on a server or backend, so no setup of this kind is necessary for running or deployment.

## Deploying

When you're ready to publish the webapp, all you need to do is build the application.

```
yarn build
```

This will create the `build/` directory. On a successful build execution, copy the `build/` directory to a web directory on your server.

## Changing the favicon

If you find yourself needing to change the favicon, export the full-resolution file to the root `favicon.png` file. After that, use the package script to generate the various favicon files.

```
yarn favicon
```

## Updating data

All of the data for the application is managed through a SQLite database for ease of use; the database allows us to have oversight in the way of foreign and unique keys while also cutting down to the bare minimum of necessary boilerplate.

The database itself is not committed to the repository because, as a binary file, resolving merge conflicts is nigh impossible. Instead, we store the SQL files for the contents and can rebuild the database at any moment.

When you're working on data, the first step will be creating your local database.

```
yarn db:create
```

This will create a local database (deleting the file that was there before, so make sure to **dump** your changes if you had any!) and will also pull in all of the changes to your database to bring it up to date.

You then modify the SQLite database `jyosuushi.sqlite` in the root directory directly. When you have made your changes, you'll then need to run the script to export the data from there into TypeScript.

```
yarn db:export
```

When this is done, it will update the files in the `./data` directory. Subsequent builds or development servers will build against these updated files. Additionally, these files are tracked in the repository to allow for painlessly building after cloning the repository.

When you are ready to submit a commit, you will need to dump all of your database back to the SQL files. To do this, you'll want to use another package script:

```
yarn db:dump
```

## Built With

- [TypeScript](https://www.typescriptlang.org/)
- [React](https://reactjs.org/) - The web framework for the application
- [Redux](https://redux.js.org/) - State management for the application
- [webpack](https://webpack.js.org/) - Module bundler and build system
- [Yarn](https://yarnpkg.com/) - Package management
- [SQLite](https://www.sqlite.org/index.html) - Pre-build database for quiz data

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/ahlec/jyosuushi/tags).

## Authors

- **Alec Deitloff** - _Initial work_ - [ahlec](https://github.com/ahlec)

See also the list of [contributors](https://github.com/ahlec/jyosuushi/contributors) who participated in this project.

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

- [WaniKani](https://www.wanikani.com/) - Inspiration and a constant guiding light
- [jquery-kana-input](https://github.com/argelius/jquery-kana-input) - Implementation basis for making a kana-based input box


CHANGELOG.md
### v2.5.3

Bug Fixes:

- [Firefox] Fixed a bug that prevented using the enter key to progress to the next question during quizzes.

### v2.5.2

Bug Fixes:

- Fixed an error that would prevent progression in quizzes that contained the counter 時 (that is, the _Essential_ and _JLPT N5_ study packs).

### v2.5.1

Bug Fixes:

- Fixed a bug where counters that used alternative readings (such as 人 allow for 4 to be read as "よ") would incorrectly apply this change elsewhere in the number (producing readings like よじゅうよにん instead of よ**ん**じゅうよにん).

Developer Changes:

- Migrated from TSLint to ESLint.

### v2.5.0

New Features:

- **New domain name.** We've moved! You can now find us at [https://www.jyosuushi.com/](https://www.jyosuushi.com/).
- **New layout for counters page.** The individual counters page have been given a new layout to help them grow to include more information. You'll start to notice some entries having links to other articles, descriptions, and comparisons with similar counters and what makes them different.
- **New counters added.** A couple more counters have joined the fray!

Developer Changes:

- Upgraded dependencies to their latest versions.
- Reworked the internal auditing scripts to speed up data development and create stronger oversight on data.
- Introduced aliasing for imports
- Improved support for linting and format enforcement.

### v2.4.0

New Features:

- **New counters added.** We have a couple more added to the pool now! Trying to ramp production up.
- **Preivew counters.** You can preview any amount for a counter now when visiting its page!

Bug Fixes:

- Fixed a bug where per-item maximum values were ignored and it lead to wonky questions (like 39 0'clock).
- Fixed a bug where you could go to pages for counters or study packs that didn't exist.

Developer Changes:

- Database is now stored in SQL files rather than a SQLite database, to improve conflict resolution and readability.
- Implemented a scraper for a website with a ton of counters, to collect items grouped by counter.

### v2.3.0

New Features:

- **New counters added.** It's slow but steady work. A number of new counters are now available!
- **Added a settings page.** You can now start changing settings around how the site works!
  - Control the range of amounts that will be generated for questions.
  - Make the quiz last forever in endless mode!
- **Further tweak amount generation.** Further work was done to make sure that the random amounts you're quizzed with will be better distributed so you get more variation, and you can focus more on interesting amounts rather than mundane.

Developer Changes:

- Added a script that's able to audit the database and detect points of interest.

### v2.2.0

New Features:

- **New layout.** The home screen has been updated! A sidebar was added, and almost everything that was previously in a modal now has its own page. With the increase in space, more features will be able to come very shortly!

Bug Fixes:

- Fixed a bug where the number of irregular conjugations after the first 17 wouldn't be included in the count of total irregulars.
- Added a missing conjugation rule between 8 and か行.

Developer Changes:

- Upgraded _all_ packages to their latest versions!
- Defined counters that do not have associated items are no longer exported from the database.
- Added `react-router` and set up the application to work as a true SPA.

### v2.1.0

New Features:

- **Improved question generation.** An attempt was made to focus more on the core numbers (1-10) and anything irregular or unexpected, and less on 42, 44, 45, 49 for every question.
- **Reduced the number of questions per counter.** It was... very high. Now it's less high.
- **Added release notes!** You'll see what changed when you load the page!

Bug Fixes:

- [Safari] Fixed a bug where the header would not fully transition.
- [Safari] Fixed a bug where the tutorial had the text rendering on top of images.
- [Firefox] Fixed a bug where the furigana were floating off to the side of the kanji.
- Switched to using Google Forms to address issues with previous bug report/submission box.

### v2.0.0

Initial release.


changelog.ts
/**
 * Ideally, we'd have the ability to sort these in a
 * fixed order:
 *  1. version
 *  2. date
 *  3. newFeatures
 *  4. improvements
 *  5. bugFixes
 *  6. developerChanges
 * So that it mirrors the order that we intend to display
 * it on the site. But `sort-keys` isn't THAT flexible and
 * it feels like it'd be more confusing to have it in
 * alphabetical order here.
 */
/* eslint-disable sort-keys */

export interface ConsumerFacingEntry {
  label: string;
  details: string;
  nestedListItems?: readonly string[];
}

export type BugFixBrowser = "chrome" | "firefox" | "safari" | "edge" | "ie";

export interface BugFixEntry {
  browser?: BugFixBrowser;
  text: string;
}

export interface IncrementalVersion {
  date: string;
  version: string;
  bugFixes: readonly BugFixEntry[];
  newFeatures: readonly ConsumerFacingEntry[];
  improvements: readonly ConsumerFacingEntry[];
  developerChanges: readonly string[];
}

export interface FirstVersion {
  version: string;
  date: string;
  isInitialRelease: true;
}

export type ChangelogVersion = Readonly<FirstVersion | IncrementalVersion>;

export function isFirstVersion(
  version: ChangelogVersion
): version is FirstVersion {
  return "isInitialRelease" in version && version.isInitialRelease === true;
}

const CHANGELOG: readonly ChangelogVersion[] = [
  {
    version: "v2.5.3",
    date: "15 July 2020",
    newFeatures: [],
    improvements: [],
    bugFixes: [
      {
        browser: "firefox",
        text:
          "Fixed a bug that prevented using the enter key to progress to the next question during quizzes.",
      },
    ],
    developerChanges: [],
  },
  {
    version: "v2.5.2",
    date: "26 May 2020",
    newFeatures: [],
    improvements: [],
    bugFixes: [
      {
        text:
          "Fixed an error that would prevent progression in quizzes that contained the counter 時 (that is, the _Essential_ and _JLPT N5_ study packs).",
      },
    ],
    developerChanges: [],
  },
  {
    version: "v2.5.1",
    date: "23 February 2020",
    newFeatures: [],
    improvements: [],
    bugFixes: [
      {
        text:
          'Fixed a bug where counters that used alternative readings (such as 人 allow for 4 to be read as "よ") would incorrectly apply this change elsewhere in the number (producing readings like よじゅうよにん instead of よ**ん**じゅうよにん).',
      },
    ],
    developerChanges: ["Migrated from TSLint to ESLint."],
  },
  {
    version: "v2.5.0",
    date: "16 February 2020",
    newFeatures: [
      {
        label: "New domain name.",
        details:
          "We've moved! You can now find us at [https://www.jyosuushi.com/](https://www.jyosuushi.com/).",
      },
      {
        label: "New counters added.",
        details: "A couple more counters have joined the fray!",
      },
    ],
    improvements: [
      {
        label: "New layout for counters page.",
        details:
          "The individual counters page have been given a new layout to help them grow to include more information. You'll start to notice some entries having links to other articles, descriptions, and comparisons with similar counters and what makes them different.",
      },
    ],
    bugFixes: [],
    developerChanges: [
      "Upgraded dependencies to their latest versions.",
      "Reworked the internal auditing scripts to speed up data development and create stronger oversight on data.",
      "Introduced aliasing for imports.",
      "Improved support for linting and format enforcement.",
    ],
  },
  {
    version: "v2.4.0",
    date: "23 September 2019",
    newFeatures: [
      {
        label: "New counters added.",
        details:
          "We have a couple more added to the pool now! Trying to ramp production up.",
      },
    ],
    improvements: [
      {
        label: "Preview counters.",
        details:
          "You can preview any amount for a counter now when visiting its page!",
      },
    ],
    bugFixes: [
      {
        text:
          "Fixed a bug where per-item maximum values were ignored and it lead to wonky questions (like 39 0'clock).",
      },
      {
        text:
          "Fixed a bug where you could go to pages for counters or study packs that didn't exist.",
      },
    ],
    developerChanges: [
      "Database is now stored in SQL files rather than a SQLite database, to improve conflict resolution and readability.",
      "Implemented a scraper for a website with a ton of counters, to collect items grouped by counter.",
    ],
  },
  {
    version: "v2.3.0",
    date: "15 July 2019",
    newFeatures: [
      {
        label: "New counters added.",
        details:
          "It's slow but steady work. A number of new counters are now available!",
      },
      {
        label: "Added a settings page.",
        details:
          "You can now start changing settings around how the site works!",
        nestedListItems: [
          "Control the range of amounts that will be generated for questions.",
          "Make the quiz last forever in endless mode!",
        ],
      },
    ],
    improvements: [
      {
        label: "Further tweak amount generation.",
        details:
          "Further work was done to make sure that the random amounts you're quizzed with will be better distributed so you get more variation, and you can focus more on interesting amounts rather than mundane.",
      },
    ],
    bugFixes: [],
    developerChanges: [
      "Added a script that's able to audit the database and detect points of interest.",
    ],
  },
  {
    version: "v2.2.0",
    date: "5 July 2019",
    newFeatures: [
      {
        label: "New layout.",
        details:
          "The home screen has been updated! A sidebar was added, and almost everything that was previously in a modal now has its own page. With the increase in space, more features will be able to come very shortly!",
      },
    ],
    improvements: [],
    bugFixes: [
      {
        text:
          "Fixed a bug where the number of irregular conjugations after the first 17 wouldn't be included in the count of total irregulars.",
      },
      { text: "Added a missing conjugation rule between 8 and か行." },
    ],
    developerChanges: [
      "Upgraded _all_ packages to their latest versions!",
      "Defined counters that do not have associated items are no longer exported from the database.",
      "Added `react-router` and set up the application to work as a true SPA.",
    ],
  },
  {
    version: "v2.1.0",
    date: "25 June 2019",
    newFeatures: [
      {
        label: "Added release notes!",
        details: "You'll see what changed when you load the page!",
      },
    ],
    improvements: [
      {
        label: "Improved question generation.",
        details:
          "An attempt was made to focus more on the core numbers (1-10) and anything irregular or unexpected, and less on 42, 44, 45, 49 for every question.",
      },
      {
        label: "Reduced the number of questions per counter.",
        details: "It was... very high. Now it's less high.",
      },
    ],
    bugFixes: [
      {
        browser: "safari",
        text: "Fixed a bug where the header would not fully transition.",
      },
      {
        browser: "safari",
        text:
          "Fixed a bug where the tutorial had the text rendering on top of images.",
      },
      {
        browser: "firefox",
        text:
          "Fixed a bug where the furigana were floating off to the side of the kanji.",
      },
      {
        text:
          "Switched to using Google Forms to address issues with previous bug report/submission box.",
      },
    ],
    developerChanges: [],
  },
  {
    version: "v2.0.0",
    date: "21 June 2019",
    isInitialRelease: true,
  },
];

export default CHANGELOG;
